srml/staking/src/slashing.rs

```diff
@@ -0,0 +1,772 @@
+pub(crate) fn clear_era_metadata<T: Trait>(obsolete_era: EraIndex) {
+	<Module<T> as Store>::ValidatorSlashInEra::remove_prefix(&obsolete_era);
+	<Module<T> as Store>::NominatorSlashInEra::remove_prefix(&obsolete_era);
+}
+
+/// Clear slashing metadata for a dead account.
+pub(crate) fn clear_stash_metadata<T: Trait>(stash: &T::AccountId) {
+	let spans = match <Module<T> as Store>::SlashingSpans::take(stash) {
+		None => return,
+		Some(s) => s,
+	};
+
+	// kill slashing-span metadata for account.
+	//
+	// this can only happen while the account is staked _if_ they are completely slashed.
+	// in that case, they may re-bond, but it would count again as span 0. Further ancient
+	// slashes would slash into this new bond, since metadata has now been cleared.
+	for span in spans.iter() {
+		<Module<T> as Store>::SpanSlash::remove(&(stash.clone(), span.index));
+	}
+}
+
+// apply the slash to a stash account, deducting any missing funds from the reward
+// payout, saturating at 0. this is mildly unfair but also an edge-case that
+// can only occur when overlapping locked funds have been slashed.
+fn apply_slash<T: Trait>(
+	stash: &T::AccountId,
+	value: BalanceOf<T>,
+	reward_payout: &mut BalanceOf<T>,
+	slash_imbalance: &mut NegativeImbalanceOf<T>,
+) {
+	let controller = match <Module<T>>::bonded(stash) {
+		None => return, // defensive: should always exist.
+		Some(c) => c,
+	};
+
+	let mut ledger = match <Module<T>>::ledger(&controller) {
+		Some(ledger) => ledger,
+		None => return, // nothing to do.
+	};
+
+	let mut value = value.min(ledger.active);
+
+	if !value.is_zero() {
+		ledger.active -= value;
+
+		// Avoid there being a dust balance left in the staking system.
+		if ledger.active < T::Currency::minimum_balance() {
+			value += ledger.active;
+			ledger.active = Zero::zero();
+		}
+
+		ledger.total -= value;
+
+		let (imbalance, missing) = T::Currency::slash(stash, value);
+		slash_imbalance.subsume(imbalance);
+
+		if !missing.is_zero() {
+			// deduct overslash from the reward payout
+			*reward_payout = reward_payout.saturating_sub(missing);
+		}
+
+		<Module<T>>::update_ledger(&controller, &ledger);
+
+		// trigger the event
+		<Module<T>>::deposit_event(
+			super::RawEvent::Slash(stash.clone(), value)
+		);
+	}
+}
+
+impl<T: Trait> RewardPayout<T> {
+	fn pay_to(&mut self, reporters: &[T::AccountId]) {
+		let (reward_payout, slashed_imbalance) = match self.0.take() {
+			None => return,
+			Some((payout, slashed_imbalance)) => {
+				if payout.is_zero() || reporters.is_empty() {
+					// nobody to pay out to or nothing to pay;
+					// just treat the whole value as slashed.
+					T::Slash::on_unbalanced(slashed_imbalance);
+					return
+				}
+				(payout, slashed_imbalance)
+			},
+		};
+
+		// take rewards out of the slashed imbalance.
+		let (mut reward_payout, mut value_slashed) = slashed_imbalance.split(reward_payout);
+
+		let per_reporter = reward_payout.peek() / (reporters.len() as u32).into();
```

Comment:
Should we add a comment that `reporters` are not empty, because we guard against this earlier? just to future-proof the code.